File: Bronze1-1/06-23-pm/unit08/blocks/blocks.java
/*
 * Blocks (Feb 2022)
 */
import java.util.*;
public class blocks {

	static int[][] perms = { //permutations of the 4 blocks
		{0,1,2,3}, {0,1,3,2}, {0,2,1,3}, {0,2,3,1}, {0,3,1,2}, {0,3,2,1},
		{1,0,2,3}, {1,0,3,2}, {1,2,0,3}, {1,2,3,0}, {1,3,0,2}, {1,3,2,0},
		{2,0,1,3}, {2,0,3,1}, {2,1,0,3}, {2,1,3,0}, {2,3,0,1}, {2,3,1,0},
		{3,0,1,2}, {3,0,2,1}, {3,1,0,2}, {3,1,2,0}, {3,2,0,1}, {3,2,1,0}
	};
	
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int N = in.nextInt(); //# of words to be spelled
		
		int[][] blocks = new int[4][26];
		for(int j=0; j<4; j++) {
			String line = in.next();
			for(int k=0; k<line.length(); k++) {
				int id = line.charAt(k)-'A';
				blocks[j][id] = 1;
			}
		}
		
		for(int j=0; j<N; j++) {
			String word = in.next();
			
			boolean success = false;
			for(int[] p : perms) {
				//can the 4 blocks in order p spell word?
				success = true;
				for(int x=0; x<word.length(); x++) {
					int id = word.charAt(x)-'A';
					if(blocks[p[x]][id]==0) {
						success = false;
						break;
					}
				}
				if(success) break;
			}
			System.out.println(success ? "YES" : "NO");
		}
		in.close();
	}
}

File: Bronze1-1/06-23-pm/unit08/blocks/blocks.cpp
/*
 * Blocks (Feb 2022)
 */
#include <bits/stdc++.h>
using namespace std;
 
bool solve(array<string, 4> blocks) {
	string word;
	cin >> word;
	do {
		bool ok = true;
		for(int i=0; i<word.size(); i++) {
			if(find(blocks[i].begin(), blocks[i].end(), word[i]) == blocks[i].end())
				ok = false;
		}
		if(ok) return true;
	} 
	while(next_permutation(blocks.begin(), blocks.end()));

	return false;
}
 
int main() {
	int TC; cin >> TC;
	array<string, 4> blocks;
	for(int i=0; i<4; i++) 
		cin >> blocks[i];
	sort(blocks.begin(), blocks.end());

	for(int i=0; i<TC; i++) {
		bool b = solve(blocks);
		if(b) 
			cout << "YES\n";
		else 
			cout << "NO\n";
	}
}
